### DISCUSSION TITLE
Access the frontmatter of all documents (at least at build time)

### DISCUSSION THREAD

**Discussion by ZenaMel** [facebook/docusaurus#9867] [Q&A] [answered]
This might go beyond Docusaurus' scope, but if this can be done within Docusaurus, it would make the solution elegant.
What I'm trying to do - our frontmatter has a field `approved` and we'd like to create a dashboard-kind-of page, similar to test coverage dashboards to keep track of which documents need to be reviewed by the team before the final release.

Something [like this](https://github.com/ezy/code-coverage-dashboard):
![image](https://github.com/facebook/docusaurus/assets/47085752/dc65ffb3-90d7-4e61-84ad-968807dfc183)

An attempt at parsing the frontmatter:

```js
    const IS_SERVER = typeof window === 'undefined';
    const IS_INTERNAL = siteConfig.customFields.ENVIRONMENT_VARIABLES.INTERNAL_BUILD === 'true';
        if (IS_SERVER) {
            // parse all markdown files (.md and .mdx) in the /docs/Controller folder
            const fs = require('fs');
            const path = require('path');
            const matter = require('gray-matter'); // NOTE: can't rely on docusausurus frontmatter parser as we can't use the appropriate context from outside markdown pages.
            const docsDir = path.join(process.cwd(), 'docs/Controller');
            const fileNames = fs.readdirSync(docsDir);
            const markdownFiles = fileNames.filter((file) => file.endsWith('.md') || file.endsWith('.mdx'));
            const markdownData = markdownFiles.map((file) => {
                const fullPath = path.join(docsDir, file);
                const fileContent = fs.readFileSync(fullPath, 'utf8');
                const { data } = matter(fileContent);
                return data;
            });
            console.log(markdownData);
```

This gives me a `Module not found: Error: Can't resolve 'fs' in 'C:\PRODUCT-DOCS\src\pages'` error as one would expect.
Is there a straight-forward way to get the frontmatter of all files at buildtime in a "docusaurus-fashion" way?

**Comment by slorber** [comment, marked as answer]
> This gives me a `Module not found: Error: Can't resolve 'fs' in 'C:\PRODUCT-DOCS\src\pages'` error as one would expect. 

🤷‍♂️ why one would expect that exactly? I'm not even sure where the code above is supposed to run, but if it's in the config file, I don't see why you'd expect that error.

> Is there a straight-forward way to get the frontmatter of all files at buildtime in a "docusaurus-fashion" way?

Maybe try to use the new `parseFrontMatter` hook?

https://docusaurus.io/blog/releases/3.1#parsefrontmatter-hook

Also, remark plugins can access front matter:

```js
function myRemarkPlugin(this: Processor): Transformer {
  return async (tree, file) => {
    console.log(file.data.frontMatter);
  };
}
```

---

I'm not sure to understand your use case so I can't really help more, but that seems out of the scope of Docusaurus yes 😅 

We build a static site, and we don't generate stats/dashboards about Markdown files.

The code I would write for that would be similar to yours.

**Reply by ZenaMel** [reply]
I got it figured out now.
I mistakenly thought that I had to fetch the markdown files' frontmatter from within the component/page, but I can simply run a build-time script that generates a json file and place that under `static`. Accessing that at run-time from within components/pages is no problem.
